File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {

     private List<String> combinations=new ArrayList<>();
     private Map<Character, String> letters = Map.of(
        '2',
        "abc",
        '3',
        "def",
        '4',
        "ghi",
        '5',
        "jkl",
        '6',
        "mno",
        '7',
        "pqrs",
        '8',
        "tuv",
        '9',
        "wxyz"
    );
    String phoneDigits;
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return combinations;
        }

        // Initiate backtracking with an empty path and starting index of 0
        phoneDigits = digits;
        backtrack(0, new StringBuilder());
        return combinations;

        
    }

    private void backtrack(int index,StringBuilder path)
    {
        if(path.length()==phoneDigits.length())
        {
            combinations.add(path.toString());
            return;
        }

        String possibleLetters=letters.get(phoneDigits.charAt(index));
        for(char letter:possibleLetters.toCharArray())
        {
            path.append(letter);
            backtrack(index+1,path);
            path.deleteCharAt(path.length()-1);
        }

    }
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
       
       vector<vector<char>> digitMap;
       digitMap.push_back({'a','b','c'});
       digitMap.push_back({'d','e','f'});
       digitMap.push_back({'g','h','i'});
       digitMap.push_back({'j','k','l'});
       digitMap.push_back({'m','n','o'});
       digitMap.push_back({'p','q','r','s'});
       digitMap.push_back({'t','u','v'});
       digitMap.push_back({'w','x','y','z'});
       int ans=0;
       if(digits.length()==0)
       {
           return {};
       }
       else
       {
           string tmp="";
           int i=0;
           vector<string> ans;
           PrintAllComb(tmp, digits, i, digitMap, ans);
           return ans;
       
    }
    }
       
       void PrintAllComb(string tmp, string &digits, int i, vector<vector<char>> &digitMap, vector<string> &ans)
       {
           if(digits[i]=='\0')
           {
           ans.push_back(tmp);
           return;
           }

           for(int j=0;j<digitMap[digits[i]-'2'].size();j++)
           {
               PrintAllComb(tmp+digitMap[digits[i]-'2'][j],digits,i+1,digitMap,ans);
           }
       }

    
};